runtime/auth/aws-signing-common/common/src/aws/smithy/kotlin/runtime/auth/awssigning/AbstractAwsChunked.kt

```diff
@@ -0,0 +1,227 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.auth.awssigning
+
+import aws.smithy.kotlin.runtime.http.Headers
+import aws.smithy.kotlin.runtime.io.SdkByteReadChannel
+
+internal abstract class AbstractAwsChunked(
+    private val chan: SdkByteReadChannel,
+    private val signer: AwsSigner,
+    private val signingConfig: AwsSigningConfig,
+    private var previousSignature: ByteArray,
+    private var trailingHeaders: Headers = Headers.Empty,
+) : SdkByteReadChannel by chan {
+    companion object {
+        const val CHUNK_SIZE_BYTES: Int = 65536
+    }
+
+    override val isClosedForRead: Boolean
+        get() = chan.isClosedForRead && (chunk == null || chunkOffset >= chunk!!.size)
+
+    var chunk: ByteArray? = null
+    var chunkOffset: Int = 0
+    var hasLastChunkBeenSent: Boolean = false
+
+    /**
+     * Returns all the bytes remaining in the underlying data source, up to [limit].
+     * @return a [ByteArray] containing at most [limit] bytes. it may contain fewer if there are less than [limit] bytes
+     * remaining in the data source.
+     */
+    override suspend fun readRemaining(limit: Int): ByteArray {
+        if (chunk == null || chunkOffset >= chunk!!.size) { chunk = getNextChunk() }
+
+        var bytesWritten = 0
+        var bytes = byteArrayOf()
+        while (bytesWritten != limit) {
+            val numBytesToWrite: Int = minOf(limit - bytesWritten, chunk!!.size - chunkOffset)
+
+            bytes += chunk!!.slice(chunkOffset until chunkOffset + numBytesToWrite)
```

Comment:
comment: this is not very efficient, can we do better than this?

Reply:
optimized this by allocating a ByteArray with `limit` size (as that's the absolute maximum we can expect to read), and truncating it before returning